pretix_eth/payment.py

```diff
@@ -223,7 +226,24 @@ def payment_control_render(self, request: HttpRequest, payment: OrderPayment):
     abort_pending_allowed = True
 
     def payment_refund_supported(self, payment: OrderPayment):
-        return False
+        return payment.state == OrderPayment.PAYMENT_STATE_CONFIRMED
 
     def payment_partial_refund_supported(self, payment: OrderPayment):
-        return False
+        return self.payment_refund_supported(payment)
+
+    def execute_refund(self, refund: OrderRefund):
+        if refund.payment is None:
+            raise ValueError("There is no associated payment with this refund")
```

Comment:
Just curious what happens at the application layer when this gets hit.  Is there something in pretix that catches this exception and displays it to the user?  Or is this an exception that we shouldn't expect to hit in production and this check is just to be sure we don't somehow execute invalid refunds?

Reply:
In the frontend pretix always creates refund object with payment:
https://github.com/pretix/pretix/blob/818d75ddd7106754922484d803910def11b726e6/src/pretix/api/views/order.py#L1110-L1116

But my concern is that, for some reason, the model allows to have payment field to be null or blank:
https://github.com/pretix/pretix/blob/818d75ddd7106754922484d803910def11b726e6/src/pretix/base/models/orders.py#L1702-L1707

Reply:
So IIUC, we don't ever expect this exception to be raised.  Instead, this is defensive to ensure that we cannot execute this method with an invalid state.  :+1: 

I'd recommend converting the exception to just be `Exception("Invariant: No payment associated with refund")`.  The use of the bare `Exception` class is a sort of signal that says, "we don't expect this to actually happen... it is just here to be sure it can't"